cpp/src/io/utilities/parsing_utils.cu

```diff
@@ -0,0 +1,143 @@
+#include "parsing_utils.cuh"
+
+#include <cuda_runtime.h>
+
+#include <vector>
+
+#include "rmm/rmm.h"
+#include "rmm/thrust_rmm_allocator.h"
+#include "utilities/error_utils.hpp"
+
+constexpr size_t max_chunk_bytes = 256*1024*1024; // 256MB
+constexpr int bytes_per_find_thread = 64;
+
+/**---------------------------------------------------------------------------*
+ * @brief Searches the input character array for each of characters in a set
+ * and sums up the number of occurences.
+ *
+ * Does not load the entire buffer into the GPU memory at any time, so it can 
+ * be used with buffers of any size.
+ *
+ * @param[in] h_data Pointer to the data in host memory
+ * @param[in] h_size Size of the input data, in bytes
+ * @param[in] keys Vector containing the keys to count in the buffer
+ * @param[out] count Total number of occurences of all keys
+ *
+ * @return gdf_error
+ *---------------------------------------------------------------------------**/
+gdf_error countAllFromSet(const char *h_data, size_t h_size, std::vector<char> keys, gdf_size_type &count)
+{
+	const size_t chunk_count = (h_size + max_chunk_bytes - 1) / max_chunk_bytes;
+
+	char* d_chunk = nullptr;
+	RMM_TRY(RMM_ALLOC (&d_chunk, min(max_chunk_bytes, h_size), 0)); 
+
+	for (size_t ci = 0; ci < chunk_count; ++ci) {
+		const auto h_chunk = h_data + ci * max_chunk_bytes;
+		const auto chunk_bytes = std::min((size_t)(h_size - ci * max_chunk_bytes), max_chunk_bytes);
+
+		// Copy chunk to device
+		CUDA_TRY(cudaMemcpyAsync(d_chunk, h_chunk, chunk_bytes, cudaMemcpyDefault));
+
+		for (char key: keys) {
+			count += thrust::count(rmm::exec_policy()->on(0), d_chunk, d_chunk + chunk_bytes, key);
```

Comment:
I'd rather we re-use the same kernel that we use for storing the start offsets. The logic for the paths to me is that we count in the first pass, then store in the second pass.

We re-use the same kernel in the Parquet reader; we launch to count the page headers, allocate memory, then launch to decode the page headers.

Reply:
Done